### PR TITLE
Update parallel training to use DDP

### PR DESCRIPTION
**Pull request by subramen**
[PyTorch DDP](https://pytorch.org/tutorials/beginner/dist_overview.html) offers better performance than DP, and is the recommended method to use for data-parallel training. This PR changes the training function to use single-machine multi-GPU DDP. DDP maximizes utilization of all GPUs; in contrast, DP is bottlenecked by the GPU that marshals data to other devices.

The PR uses `torch.utils.data.distributed.DistributedSampler` to divide training and eval batches across devices, and `torch.multiprocessing` to spawn one process per GPU. Each process initializes its own replica of the model, dataloaders and optimizer. Gradients are synced across processes at every backward pass.

(!) A limitation of multiprocessing is that it cannot be directly used in ipython; the notebook cell that contains the multiprocessing code (`mp.spawn(train_model, nprocs=ngpus, args=(ngpus, vocab_src, vocab_tgt, *train_args.values()))`) throws an AttributeError. One way around this is to define the spawned function in a separate file, but I am not sure how suitable that is for our purposes.

@austinvhuang

### CODE REVIEW

**Comment by srush** [comment]
looks cool, but need to run style checks.

**Comment by austinvhuang** [comment]
Thanks @suraj813 !

With regard to being able to import functions, can you use the fact that there's already a python `.py` script with all the functions paired with the `.ipynb` notebook since we're using jupytext?

If it makes it easier, you can `git mv` `The Annotated Transformer.py` to a file name that's more import friendly like `the_annotated_transformer.py` (changing the `makefile` and readme accordingly).

Definitely kind of awkward though, we can explore other options if that doesn't work.

**Comment by subramen** [comment]
Thanks @austinvhuang, importing the function from the source script works, but with the (slightly awkward) addition of an `IS_NOTEBOOK` flag to avoid unnecessary execution when importing. Please let me know if there is a better way to do this?

**Comment by austinvhuang** [comment]
@suraj813 nice that it's working - will have a closer look tonight.

**Comment by austinvhuang** [comment]
A solution to cleaning up the IS_NOTEBOOK guards is to do something we needed to do as a cleanup pass anyway -use show_example() guards around execution consistently for pretty much all the cells with execution. Then we'll have just the one check inside show_example. I think this should work but I haven't tried it so there might be a little fiddling:

- The entire creation of vocab_src, vocab_tgt [1446-1455](https://github.com/suraj813/annotated-transformer/blob/d3f7ac4b79876674513b6aa92cc64aaace97dcfb/the_annotated_transformer.py#L1446) could be wrapped in a function, which is then called by show_example, with the vocab_src, vocab_tgt globals being returned as a tuple for use in later contexts. 

- I think you might be able to do the same with [train_model](https://github.com/suraj813/annotated-transformer/blob/d3f7ac4b79876674513b6aa92cc64aaace97dcfb/the_annotated_transformer.py#L1722)

- [Here](https://github.com/suraj813/annotated-transformer/blob/d3f7ac4b79876674513b6aa92cc64aaace97dcfb/the_annotated_transformer.py#L1887) again I think the executed snippet with the check_outputs call could be wrapped in show_example.

- Same thing with the [last few cells plotting attention maps](https://github.com/suraj813/annotated-transformer/blob/d3f7ac4b79876674513b6aa92cc64aaace97dcfb/the_annotated_transformer.py#L1991), can wrap them in functions and wrap the function in show_example. 

If this works I can go ahead and merge it. Thanks!

**Comment by subramen** [comment]
I ran into some issues while refactoring, I'll take another stab at it and have it ready for merging in a couple days

**Comment by subramen** [comment]
@austinvhuang  I had to refactor some functions that used global variables because they didn't work correctly in an `mp` context. Getting multiprocessing to play well with ipynb was tricky, but now it's working as expected! 

I also took this chance to refactor the training loop to allow non-multi GPU training, to show what code for that looks like.

**Comment by austinvhuang** [comment]
LGTM will go ahead and merge. Will try to find time this weekend to make sure I'm able to run it without any issues.

**Comment by austinvhuang** [comment]
@suraj813 taking a stab at reproducing the run on my end. Seem to be running into some problematic interaction between torchtext's data downloading and multiprocessing. Did you come across this?

image

**Comment by austinvhuang** [comment]
Might be related to this:

https://github.com/pytorch/text/issues/1359
and this from just yesterday
https://github.com/pytorch/text/issues/1620

which are open issues. I'm not sure why we never ran into these before though.

**Comment by subramen** [comment]
I don't see this error on AWS (torchtext==0.11.0), but I am running into it on my local machine (tested with v0.11.0  and v0.11.2). I'll try following up on the open issues internally ✌️
